POSIX source compatibility and clang with Apple

ChisholmKyle opened the pull request:
I'm not sure if this should all be one merge, but I was working on sources that define _POSIX_C_SOURCE which causes some issues with the header. I tested the changes on Debian 8 and Mac OSX. Please let me know what you think. Thanks.

ChisholmKyle commented:
I've also added a few string assertions. I'm not sure if this should be a feature request or an issue before creating a pull request. Either way, I like this simple unit test framework you have here :)

bvdberg commented:
I like your changes, but they don't build under normal Ubuntu 16:
ctest.h:422:45: error: ‘sys_siglist’ undeclared (first use in this function)
     sprintf(msg, "[SIGNAL %d: %s]", signum, sys_siglist[signum]);

I also think your string compares are a good extension, but I why the complex NULL checking?
If either string is NULL, there should be an error right?

ChisholmKyle commented:
It appears the function `sys_siglist` is a BSD library function and is also implemented with GNU extensions. I added some checks for feature test macros: `_BSD_SOURCE` or `_POSIX_C_SOURCE` or `_GNU_SOURCE` should be defined for the `CTEST_SEGFAULT` feature. The easiest thing to do is compile with `-std=gnu99` (which is default for gcc these days I think). The `kill` function is also a POSIX or GNU extension. All these extensions to the standard c library are annoying but if anyone wants to just use the c99 standard (with `-std=c99`) then I think these extensions should be explicitly checked. For the example, using `-std=gun99` makes sense to enable the features but should also compile with `std=c99`

Also changed the test to check for the functions `usleep` or `nanosleep` which are not standard C99. 

You're right about all the NULL checks. I think it was to make sure NULL was not passed to `strcmp` or `strlen` (which is not allowed/undefined) but that is taken care of with the pointer checks `exp && real && ...`
I made the changes and committed. Any other thoughts?

ChisholmKyle commented:
I haven't really thought about windows compatibility. Maybe worth checking?

bvdberg commented:
I don't want to couple ctest to gcc. I'm using clang a lot nowadays, so std=c99 would be an option.

I'm not a Windows developer myself, so I cannot test that part. If you could it would be nice, but otherwise I don't mind just merging. Could you please squash your changes? because now everyone sees all the commits..

Bas

ChisholmKyle commented:
Okay cleaned up all those commits and set the compiler flag `-std=c99` and used nanosleep instead of usleep.

mikedld commented on ctest.h:
```diff
@@ -16,6 +16,10 @@
 #define CTEST_H
 
+#if !(defined (_GNU_SOURCE) || defined (_BSD_SOURCE) || defined(_POSIX_C_SOURCE))
+#undef CTEST_SEGFAULT
```
Not sure why this is necessary. While `kill` isn't standard, in the particular case below it could be substituted with `raise(signum);` which is standard alright. Am I missing something?

bvdberg replied:
That could be a good idea. From the documentation, kill() should also be present all almost all modern systems. Chris, did you run into a problem with this or is your proposal to make it more correct?

mikedld replied:
Windows isn't one of those modern systems anyway, so if we want to support it in the future then `kill` must go (but not by disabling the whole segfault feature) IMHO :)

bvdberg replied:
On my list, Linux, Bsd and Mac are all target systems. Windows comes trailing behind by a mile or so. Nice if we can support with with minor changes, but I don't want to have to go to great lengths for it..

mikedld commented on ctest.h:
```diff
@@ -391,7 +419,7 @@ static void color_print(const char* color, const char* text) {
 }
 
-#ifdef __APPLE__
+#if defined (__APPLE__) && !defined (_POSIX_C_SOURCE)
```
So with POSIX compatibility we lose setup and teardown on Mac? If we don't, is this code necessary at all?

ChisholmKyle replied:
It appears `RTLD_DEFAULT` is only defined if the feature test macro `_GNU_SOURCE` is defined in the source when compiling with `-std=c99`. If both `_GNU_SOURCE` and `_C_POSIX_SOURCE` are defined, it appears the POSIX overrides the GNU extensions that defines `RTLD_DEFAULT` (and also the function `sys_siglist`).

ChisholmKyle replied:
I'm not sure if it's the same on Linux. On that note, why is setup/teardown feature for Apple only? I don't see Apple-only functions in `find_symbol`

bvdberg replied:
setup/teardown is definately not an Apple-only feature! However, for Apple we need to do an extra step, the find_symbol() thing..

ChisholmKyle commented:
What do you want to do with this pull request? If you don't want to check feature macros that it may be left up to the programmer to ensure they are compiling with the correct flags and/or macros to enable the features ctest depends on (kill, sys_siglist, etc.). If not explicitly checked to disable some functionality like `CTEST_SEGFAULT`, the specific required extensions to the C language (extensions such as BSD, GNU or POSIX) should be properly documented as requirements to compile ctest. However, it would be nice to be able to compile with strictly standard c99 or only POSIX, for example, even if some features are disabled, which is why the tests were added to this pull request.

ChisholmKyle commented:
I haven't used ctest in a while and this request has gone stagnant so I'll just close it.
